### ISSUE TITLE
Random data to curve points

### ISSUE THREAD

**Issue by bellebaum** [relic-toolkit/relic#307]
I am trying to implement [BBS signatures](https://datatracker.ietf.org/doc/draft-irtf-cfrg-bbs-signatures/), for which I need a Hash-to-Curve for BLS12-381.
I basically need something similar to the following function:

https://github.com/relic-toolkit/relic/blob/3b9a4a8ab2b854953d9982040fa1ca74f57c06c7/src/ep/relic_ep_map.c#L208-L232

Except, I need to be able to exchange the DST for `md_xmd`, which is currently hardcoded to "RELIC", and for some modes, replace the entire XOF with SHAKE.
My current workaround is to reimplement much of `src/ep/relic_ep_map.c` and include some internal template headers.

Would it be possible to provide an interface publicly which takes random data of the correct length (i.e. `r`) instead of deriving it from a message?

**Comment by dfaranha** [comment]
Take a look at 885808418057f236a0484c5dab32d34c2f614fbc, feedback is welcome!

I just need to add checks to ensure enough randomness is given, but the API should stay the same.

**Comment by bellebaum** [comment]
Thanks :)
I left a few comments on the commit. As you mentioned, checks on `len` are probably a good idea

**Comment by dfaranha** [comment]
I just pushed some error-handling to HEAD, it should be better now.

**Comment by bellebaum** [comment]
Thanks :)
Though it still seems a bit weird to require `2*elm+1` bytes even though only one map needs that many:

https://github.com/relic-toolkit/relic/blob/0cff31bb9e157ab9c81b1c7ac38c3cbf29cccb2d/src/ep/relic_ep_map.c#L547-L554

Maybe guards in `ep_map_rnd_size` would make sense, something like this:

```c
size_t ep_map_rnd_size(void) {
	const size_t elm = (FP_PRIME + ep_param_level() + 7) / 8;


#if EP_MAP == BASIC || !defined(STRIP)
	return elm;
#elif EP_MAP == SSWUM || !defined(STRIP)
	return 2 * elm;
#elif EP_MAP == SWIFT || !defined(STRIP)
	return 2 * elm + 1;
#endif
}
```

This would work as long as `STRIP` is defined.
I am not entirely sure what not defining `STRIP` is supposed to do. My initial idea was that defining it should have an effect like `-ffunction-sections -Wl,--gc-sections` except not even compiling the unneeded functions, but `ep_rnd_map` right now would calculate all three maps if `STRIP` is undefined.

**Comment by dfaranha** [comment]
Oh, that with `STRIP` is a bug, fixing ASAP.

The problem with requiring less bytes is obtaining from the API a lower number of bytes than the chosen function needs due to the different configurations.

**Comment by bellebaum** [comment]
Can you clarify what you mean by different configurations? `ep_map_rnd_size` and `ep_map_rnd` live within the same translation unit, so however `EP_MAP` is defined, an array of `ep_map_rnd_size()` random bytes should be sufficient for a call to `ep_map_rnd`, I would think

**Comment by dfaranha** [comment]
Yeah, I'm worried about someone querying `ep_map_rnd_size()` and supplying that amount of data to one of the internal functions, but I'm probably overthinking this. If everything goes through the API, it's clean.

I'll commit and push then.

**Comment by dfaranha** [comment]
Done!

**Comment by bellebaum** [comment]
Nice. I'll give it a try in our BBS project and report any problems I come across

**Comment by bellebaum** [comment]
Everything seems to work 👍 
Thanks again for your help and dedication. From my viewpoint, this issue can be closed

**Comment by dfaranha** [comment]
Thank you for the feedback and patience!
